pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,665 @@
+	// if object is not found, it means the object has been deleted. Log the deletion and do not requeue
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// detachment request
+	if requestedToBeDeleted(azVolumeAttachment) {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.Attached || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed || azVolumeAttachment.Status.State == v1alpha1.DetachmentFailed {
+			if err := r.triggerDetach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// attachment request
+	} else if azVolumeAttachment.Status.Detail == nil {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed {
+			if err := r.triggerAttach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// promotion request
+	} else if azVolumeAttachment.Spec.RequestedRole != azVolumeAttachment.Status.Detail.Role {
+		if err := r.promote(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, attachmentName string, azVolumeAttachment *v1alpha1.AzVolumeAttachment) error {
+	// initialize metadata and update status block
+	conditionFunc := func() (bool, error) {
+		azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, true)
+		if err != nil {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, err)
+			// if failed to get azVolumeAttachment then escape the wait
+			return false, err
+		}
+
+		// deepCopy
+		azVolumeAttachment = azVolumeAttachment.DeepCopy()
+		// Initialize finalizer and add label to the object
+		azVolumeAttachment = r.initializeMeta(ctx, azVolumeAttachment)
+		// Update state to attaching
+		azVolumeAttachment = r.updateState(ctx, azVolumeAttachment, v1alpha1.Attaching)
+
+		// update CRI
+		if err := r.update(ctx, azVolumeAttachment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
+		return err
+	}
+
+	// TODO reassess if adding additional finalizer to AzVolume CRI is necessary
+	// // add finalizer to the bound AzVolume CRI
+	// conditionFunc = func() (bool, error) {
+	// 	if err := r.addFinalizerToAzVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+	// 		return false, nil
+	// 	}
+	// 	return true, nil
+	// }
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
```

Comment:
Is this second polling-wait necessary if the code above is commented out?

Reply:
Oops, I missed that line